src/fiber/network.rs

```diff
@@ -2061,30 +2083,273 @@ where
+                        self.update_with_tcl_fail(&detail_error).await;
+                        if payment_session.can_retry() && !detail_error.error_code.payment_failed()
+                        {
+                            let res = self.try_payment_session(state, payment_session).await;
+                            if res.is_err() {
+                                debug!("Failed to retry payment session: {:?}", res);
+                            }
+                        } else {
+                            payment_session.set_failed_status(detail_error.error_code.as_ref());
+                            self.store.insert_payment_session(payment_session);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn update_with_tcl_fail(&self, tcl_error_detail: &TlcErr) {
+        let error_code = tcl_error_detail.error_code();
+        // https://github.com/lightning/bolts/blob/master/04-onion-routing.md#rationale-6
+        // we now still update the graph, maybe we need to remove it later?
+        if error_code.is_update() {
+            if let Some(extra_data) = &tcl_error_detail.extra_data {
+                match extra_data {
+                    TlcErrData::ChannelFailed { channel_update, .. } => {
+                        if let Some(channel_update) = channel_update {
+                            let mut graph = self.network_graph.write().await;
+                            let _ = graph.process_channel_update(channel_update.clone());
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+        match tcl_error_detail.error_code() {
+            TlcErrorCode::PermanentChannelFailure
+            | TlcErrorCode::ChannelDisabled
+            | TlcErrorCode::UnknownNextPeer => {
+                let channel_outpoint = tcl_error_detail
+                    .error_channel_outpoint()
+                    .expect("expect channel outpoint");
+                debug!("mark channel failed: {:?}", channel_outpoint);
+                let mut graph = self.network_graph.write().await;
+                graph.mark_channel_failed(&channel_outpoint);
+            }
+            TlcErrorCode::PermanentNodeFailure => {
+                let node_id = tcl_error_detail.error_node_id().expect("expect node id");
+                let mut graph = self.network_graph.write().await;
+                graph.mark_node_failed(node_id);
+            }
+            _ => {}
+        }
+    }
+
+    fn on_get_payment(&self, payment_hash: &Hash256) -> Result<SendPaymentResponse, Error> {
+        match self.store.get_payment_session(*payment_hash) {
+            Some(payment_session) => Ok(payment_session.into()),
+            None => Err(Error::InvalidParameter(format!(
+                "Payment session not found: {:?}",
+                payment_hash
+            ))),
+        }
+    }
+
+    async fn try_payment_session(
```

Comment:
suggest to extract two helper functions `build_route` and `send_onion_packet` to simplifying the main function.

and setting final error to `None`` outside the loop and then assigning it within the loop lead to unnecessary state management and logic complexity.

```
async fn try_payment_session(
    &self,
    state: &mut NetworkActorState<S>,
    mut payment_session: PaymentSession,
) -> Result<PaymentSession, Error> {
    let payment_data = payment_session.request.clone();
    let payment_hash = payment_data.payment_hash;

    while payment_session.can_retry() {
        payment_session.retried_times += 1;

        let hops_infos = self.build_route(payment_data.clone()).await?;
        let first_channel_outpoint = hops_infos[0]
            .channel_outpoint
            .clone()
            .expect("first hop channel outpoint");

        let tlc_id = self.send_onion_packet(state, hops_infos).await?;

        payment_session.set_status(PaymentSessionStatus::Inflight);
        payment_session.set_first_hop_info(first_channel_outpoint, tlc_id);
        self.store.insert_payment_session(payment_session.clone());

        return Ok(payment_session);
    }

    let error_msg = format!("Exhausted retries for payment: {:?}", payment_hash);
    payment_session.set_status(PaymentSessionStatus::Failed);
    payment_session.set_failed_status(&error_msg);
    self.store.insert_payment_session(payment_session);
    Err(Error::SendPaymentError(error_msg))
}
```

Reply:
When there is any error happened in `build_route` or `send_onion_packet`, instead of just throwing an exception to upper level, we need to set payment session status and store into db. 

let me think how to make the code simpler.